File: squawk/mapper/preprocessed/com/sun/squawk/suiteconverter/Suite.java
/*
 * Copyright 2005-2008 Sun Microsystems, Inc. All Rights Reserved.
 * Copyright 2011-2013 Oracle Corporation. All Rights Reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER
 *
 * This code is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2
 * only, as published by the Free Software Foundation.
 *
 * This code is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License version 2 for more details (a copy is
 * included in the LICENSE file that accompanied this code).
 *
 * You should have received a copy of the GNU General Public License
 * version 2 along with this work; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
 * 02110-1301 USA
 *
 * Please contact Oracle Corporation, 500 Oracle Parkway, Redwood
 * Shores, CA 94065 or visit www.oracle.com if you need additional
 * information or have any questions.
 */

package com.sun.squawk.suiteconverter;

import com.sun.squawk.ObjectMemoryFile;
import com.sun.squawk.ObjectMemorySerializer;
import com.sun.squawk.VM;
import java.io.DataInputStream;
import java.io.DataOutput;
import java.io.DataOutputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;

//import com.sun.spot.peripheral.ConfigPage;

//import com.sun.squawk.security.verifier.SignatureVerifier;


/**
 * Suite allows suite files to be saved in a form that can execute directly from flash memory.
 *
 * NOTE: This is a file that originally came from the spot build process.
 *
 * Squawk suites are normally saved in "canonical" form, where the base address of the bootstrap suite is assumed
 * to be 0, the base address of child suites to be directly above the last word of the bootstrap suite, and so on
 * for grandchildren. The suites also containing an OOP map that identifies which words are actually pointers. On the
 * desktop, when a suite is loaded into memory, its pointers can then be remapped according to the actual address
 * at which it finds itself, its parent, grandparent, and so on.
 *
 * On the SPOT, this remapping can't happen, as the suites are in flash memory. So instead, we remap them on the host
 * before downloading to the device. This relies on the suites living at known addresses in SPOT virtual memory.
 *
 * This class also supports a truncated suite header for the bootstrap suite, which only contains 12 bytes (no parent fields, oopmap, etc).
 *
 */
public class Suite {
    public final static int FLASH_SUITE_MINOR_VERSION = 0;
    public final static int FLASH_SUITE_MAJOR_VERSION = 2;

	private String parentURL;
	private int rootOffset;
	private int memorySize;
	private byte[] oopMap;
	private byte[] objectMemory;
	private int canonicalStart;
	private Suite parentSuite = null;
	private int outputHeaderSize;
	private short version_minor;
	private short version_major;
	private int unpaddedHdrSize;
	private int hash;

    private static boolean isBigEndian = false; // ARM target is little endian

	public static final int EXPECTED_SUITE_VERSION_MAJOR = ObjectMemorySerializer.CURRENT_MAJOR_VERSION;
    public static final int EXPECTED_SUITE_VERSION_MINOR = ObjectMemorySerializer.CURRENT_MINOR_VERSION;;

    /* copied from ConfigPage: */
    public static final String LIBRARY_URI =  "spotsuite" + "://library";

    public static boolean isTargetBigEndian() {
        return isBigEndian;
    }

    public static void setIsTargetBigEndian(boolean value) {
        isBigEndian = value;
    }

	/**
	 * Load a suite file from a {@link DataInputStream}. See also {@link #loadFromFile(String, String, int[])}
	 *
	 * @param dis The {@link DataInputStream} to read the suite from
	 * @param bootstrapFilename A filepath to read the bootstrap suite from (note that this is NOT typically
	 * this suite's parent.
	 * @throws IOException
	 */
	public void loadFromStream(DataInputStream dis, String bootstrapFilename) throws IOException {
		int magic = dis.readInt();
		if (magic != 0xDEADBEEF) {
			throw new IOException("Suite file has wrong magic word: " + Integer.toHexString(magic));
		}
		version_minor = dis.readShort();
		version_major = dis.readShort();
		if ((version_major != EXPECTED_SUITE_VERSION_MAJOR)
				|| (version_minor != EXPECTED_SUITE_VERSION_MINOR))
			throw new RuntimeException("Unsupported suite file version: "
					+ version_major + "." + version_minor + ". Expected "
					+ EXPECTED_SUITE_VERSION_MAJOR + "."
					+ EXPECTED_SUITE_VERSION_MINOR);

		dis.readInt(); // attributes
		readParentHash(dis);
		parentURL = dis.readUTF();
		rootOffset = dis.readInt();
		memorySize = dis.readInt();

		int oopMapSize = calculateOopMapSizeInBytes(memorySize);
		oopMap = new byte[oopMapSize];
		dis.readFully(oopMap);

		if (hasParent()) {
			parentSuite = new Suite();
			if (parentURL.equals("memory:bootstrap")) {
				parentSuite.loadFromFile(bootstrapFilename, null);
			} else {
				// skip "file://"
				// System.out.println(parentURL);
			    // EAT:
			    // Using the path to the bootstrap is not really the right way to go, we should have a mechanism for setting
			    // the path to use to find suites when we need them.  This used to work as the suite written out by SuiteCreator
			    // would have a full path to the original suite.  I didn't think that was a good idea and therefore introduced
			    // concept of suite class path type of thing.  In order to adapt your concept of Suite to fit this, I would have
			    // to have changed a lot of classes, and wanted to limit scope of changes I was making.
			    String parentFileName = new File(new File(bootstrapFilename).getParent(), new File(parentURL.substring("file://".length())).getName()).getAbsolutePath();
				parentSuite.loadFromFile(parentFileName, bootstrapFilename);
			}
			canonicalStart = parentSuite.getCanonicalEnd();
			unpaddedHdrSize = 24 + 2 + getSpotParentURL().length() + oopMapSize;
			outputHeaderSize = ((unpaddedHdrSize + 3) / 4) * 4;
		} else {
			outputHeaderSize = 12;
		}

		// skip padding
		dis.skipBytes((4 - ((oopMapSize + parentURL.length() + 2) % 4)) % 4);
		objectMemory = new byte[memorySize];
		dis.readFully(objectMemory);
		hash = objectMemory.length;
		for (int i = 0; i != objectMemory.length; ++i) {
			hash += objectMemory[i];
		}
        
        //logHeader("stream");
	}

    public void logHeader(String name) {
        System.out.println("SuiteReloInfo: " + name);
        System.out.println("canonicalStart: 0x" + Integer.toHexString(canonicalStart));
        System.out.println("canonicalEnd: 0x" + Integer.toHexString(canonicalStart + memorySize));
        System.out.println("parentURL: " + parentURL);
        System.out.println("rootOffset: 0x" + Integer.toHexString(rootOffset));
        System.out.println("size: " + memorySize);
        System.out.println("unpadded header size: " + unpaddedHdrSize);
        System.out.println("header size: " + ((unpaddedHdrSize + 3) / 4));
    }

	private int readParentHash(DataInputStream dis) throws IOException {
		return dis.readInt();
	}

	/**
	 * Write the suite to a stream
	 * @param dos the data output stream
	 * @throws IOException
	 */
	public void writeToStream(DataOutputStream dos) throws IOException {
///*if[!SIMPLE_VERIFY_SIGNATURES]*/
        DataOutputStream output = dos;
        final int MAX_HEADER_SIZE = Integer.MAX_VALUE;
///*else[SIMPLE_VERIFY_SIGNATURES]*/
//      SigningOutputStream output = new SigningOutputStream(dos);
//      final int MAX_HEADER_SIZE = SignatureVerifier.MAXIMUM_HEADER_SIZE;
///*end[SIMPLE_VERIFY_SIGNATURES]*/

        if (hasParent()) {
            if (VM.isVerbose()) {
                System.out.println("Converting library/app suite:");
            }
            /*
             * u4 magic // 0xDEADBEEF
             * u2 minor_version;
             * u2 major_version;
             * u4 attributes; // mask of the ATTRIBUTE_* constants in this class
             * u4 parent_hash;
             * utf8 parent_url;
             * u4 root; // offset (in bytes) in 'memory' of the root of the graph
             * u4 size; // size (in bytes) of memory  'memory' on a word boundary
             * u1 memory[size];
             */
            output.writeInt(0xDEADBEEF);
            output.writeShort(FLASH_SUITE_MINOR_VERSION); // minor
            output.writeShort(FLASH_SUITE_MAJOR_VERSION); // major
            output.writeInt(ObjectMemoryFile.ATTRIBUTE_32BIT | (isTargetBigEndian() ? ObjectMemoryFile.ATTRIBUTE_BIGENDIAN : 0)); // 32 bit
            output.writeInt(getParent().getHash());
            output.writeUTF(getSpotParentURL());
            output.writeInt(rootOffset);
            output.writeInt(memorySize);
            output.write(oopMap);
            writePad(output, outputHeaderSize - unpaddedHdrSize);
            if ((outputHeaderSize - oopMap.length) > MAX_HEADER_SIZE) {
                throw new RuntimeException(
                        "Header size of suite is too large. For compatibility with SuiteSignatureVerifier header size "
                        + "must not be larger than "
                        + MAX_HEADER_SIZE
                        + " bytes, but it is "
                        + outputHeaderSize
                        + " bytes");
            }
        } else {
            if (VM.isVerbose()) {
                System.out.println("Converting boostrap suite:");
                System.out.println("rootOffset: 0x" + Integer.toHexString(rootOffset + outputHeaderSize));
                System.out.println("hash: 0x" + Integer.toHexString(getHash()));
                System.out.println("size: 0x" + Integer.toHexString(memorySize));
            }
            writeLittleEndianInt(output, rootOffset + outputHeaderSize);
            writeLittleEndianInt(output, getHash());
            writeLittleEndianInt(output, memorySize);
        }
        output.write(objectMemory);

///*if[!SIMPLE_VERIFY_SIGNATURES]*/
        output.flush();
///*else[SIMPLE_VERIFY_SIGNATURES]*/
//		if (hasParent())
//		// If this is not the bootstrap suite write the hash
//		// and sign the suite.
//		{
//			output.writeInt(getHash());
//			output.flushAndAppendSignature();
//		}else{
//			output.flushWithoutSignature();
//		}
///*end[SIMPLE_VERIFY_SIGNATURES]*/
	}

	/**
	 * Answer a parent URL for use on the SPOT bootstrap's parent is null - same
	 * as the host lib's parent is memory:bootstrap - same as the host apps's
	 * parent (all other cases!) is different and defined as flash://<lib addr>
	 */
	private String getSpotParentURL() {
		if (parentURL.equals("memory:bootstrap") || (parentURL == null)) {
			return parentURL;
		} else {
//			return ConfigPage.LIBRARY_URI;
			return LIBRARY_URI;
		}
	}

	/**
	 * @return The calculated hash of this Suite.
	 */
	public int getHash() {
		return hash;
	}

	/**
	 * @param dos
	 * @param i
	 * @throws IOException
	 */
	private void writeLittleEndianInt(DataOutput dos, int i)
			throws IOException {
		dos.write((i >>> 0) & 0xFF);
		dos.write((i >>> 8) & 0xFF);
		dos.write((i >>> 16) & 0xFF);
		dos.write((i >>> 24) & 0xFF);
	}

	/**
	 * @param
	 * @throws IOException
	 *
	 */
	private void writePad(DataOutput dos, int padCount)
			throws IOException {
		for (int i = 0; i < padCount; i++) {
			dos.writeByte(0);
		}
	}

	/**
	 * Relocate the suite's object memory in accordance with the memory addresses supplied
	 * @param memoryAddrs  An array of virtual memory addresses at which to assume suites will be mapped on the SPOT
	 * device. The first will be this suite's address, the second its parent, and so on until the the bootstrap address.
	 */
	public void relocateMemory(int[] memoryAddrs) {
		byte[] result = new byte[objectMemory.length];
		System.arraycopy(objectMemory, 0, result, 0, objectMemory.length);
		for (int i = 0; i < oopMap.length; i++) {
			byte currentByte = oopMap[i];
			for (int j = 0; j < 8; j++) {
				if (((currentByte >> j) & 1) == 1) {
					int index = 4 * ((i * 8) + j);
					int pointer = getObjectMemoryWord(objectMemory, index);
					writeObjectMemoryWord(result, index, this.mapPointer(
							pointer, memoryAddrs));
				}
			}
		}
		objectMemory = result;
	}

	/**
	 * pointer is canonical. If it points into our memory space, then assume our
	 * flash memory base is the first element of memoryAddrs and adjust
	 * accordingly. If it doesn't then delegate to our parent, having stripped
	 * our flash memory base from the array.
	 *
	 * @param pointer value to relocate
	 * @param memoryAddrs array of actual base addresses for this suite, and for each parent suite.
	 * @param offset offset to the memory address (in memoryAddrs) that goes with "this" suite
	 * @return
	 */
	private int mapPointer(int pointer, int[] memoryAddrs, int offset) {
		if (pointer == 0) {
			return 0;
		} else {
			if (pointer < canonicalStart) {
				// map pointer against parent memory address
				return getParent().mapPointer(pointer, memoryAddrs, offset + 1);
			} else {
				// map pointer against our memory address
				return memoryAddrs[offset] - canonicalStart + pointer
						+ outputHeaderSize;
			}
		}
	}

    /**
	 * pointer is canonical. If it points into our memory space, then assume our
	 * flash memory base is the first element of memoryAddrs and adjust
	 * accordingly. If it doesn't then delegate to our parent, having stripped
	 * our flash memory base from the array.
	 *
	 * @param pointer
	 * @param parentMemoryAddrs
	 * @return
	 */
	private int mapPointer(int pointer, int[] memoryAddrs) {
        return mapPointer(pointer, memoryAddrs, 0);
	}

	/**
	 * @param result
	 * @param index
	 * @param replacementPointer
	 */
	private void writeObjectMemoryWord(byte[] memory, int index, int value) {
		if (isTargetBigEndian()) {
			memory[index + 0] = (byte) (value >> 24);
			memory[index + 1] = (byte) (value >> 16);
			memory[index + 2] = (byte) (value >> 8);
			memory[index + 3] = (byte) (value >> 0);
		} else {
			memory[index + 0] = (byte) (value >> 0);
			memory[index + 1] = (byte) (value >> 8);
			memory[index + 2] = (byte) (value >> 16);
			memory[index + 3] = (byte) (value >> 24);
		}

	}

	static private int getObjectMemoryWord(byte[] memory, int index) {
		int b0 = memory[index + 0] & 0xFF;
		int b1 = memory[index + 1] & 0xFF;
		int b2 = memory[index + 2] & 0xFF;
		int b3 = memory[index + 3] & 0xFF;
		if (isTargetBigEndian()) {
			return (b0 << 24) | (b1 << 16) | (b2 << 8) | b3;
		} else {
			return (b3 << 24) | (b2 << 16) | (b1 << 8) | b0;
		}
	}

	private int getCanonicalEnd() {
		return memorySize + canonicalStart;
	}

	private boolean hasParent() {
		return parentURL.length() != 0;
	}

	/**
	 * @return The loaded parent suite of this Suite
	 */
	public Suite getParent() {
		return parentSuite;
	}

	/**
	 * Load a suite file from a {@link File}. See also {@link #loadFromStream(DataInputStream, String)}
	 *
	 * @param filename The filepath to read the suite from
	 * @param bootstrapFilename A filepath to read the bootstrap suite from (note that this is NOT typically
	 * this suite's parent.
	 * @throws IOException
	 */
	public void loadFromFile(String filename, String bootstrapFilename) throws IOException {
		File inputFile = new File(filename);
		FileInputStream fis = new FileInputStream(inputFile);
		loadFromStream(new DataInputStream(fis), bootstrapFilename);
		fis.close();
        //logHeader(filename);
	}

	private int calculateOopMapSizeInBytes(int size) {
		return ((size / 4) + 7) / 8;
	}

}

File: squawk/vmcore/src/vm/suite.c
/*
 * Copyright 2004-2008 Sun Microsystems, Inc. All Rights Reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER
 * 
 * This code is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2
 * only, as published by the Free Software Foundation.
 * 
 * This code is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License version 2 for more details (a copy is
 * included in the LICENSE file that accompanied this code).
 * 
 * You should have received a copy of the GNU General Public License
 * version 2 along with this work; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
 * 02110-1301 USA
 * 
 * Please contact Sun Microsystems, Inc., 16 Network Circle, Menlo
 * Park, CA 94025 or visit www.sun.com if you need additional
 * information or have any questions.
 */

/**
 * This file contains functions for loading the bootstrap suite from a file and relocating the
 * contained object memory. The structure of an object memory file is described in the javadoc
 * comment of the java.lang.ObjectMemory class.
 */

// Set to true to enable tracing of the bootstrap suite file as it's read
#define TRACE_SUITE false

#ifdef FLASH_MEMORY

// the next definition needs to be kept in sync with suite converter
#define NUMBER_OF_BYTES_IN_BYTECODE_HEADER (3 * sizeof(UWord))

UWord loadBootstrapSuiteFromFlash(
		char *bootstrapSuiteFile,
		Address *romStart,
		Address *suite,
		int *hash) {
	// ROM starts at the flash address set on command line
	Address javabytecodesbase = (Address) atoi(bootstrapSuiteFile);
	*suite = (void *)(getUWord(javabytecodesbase, 0) + javabytecodesbase);
	*hash = (int)getUWord(javabytecodesbase, 1);
	UWord size=getUWord(javabytecodesbase, 2);
	*romStart=(void *)(javabytecodesbase + NUMBER_OF_BYTES_IN_BYTECODE_HEADER);

	diagnosticWithValue("javabytecodesbase: ", (int) javabytecodesbase);
	diagnosticWithValue("rootOffset: ", (int) *suite);
	diagnosticWithValue("hash: ", getUWord(javabytecodesbase, 0));
	diagnosticWithValue("suite size: ", size);
	diagnosticWithValue("romStart: ", (int) *romStart);
	return size;
}

#else /* FLASH_MEMORY */

/**
 * Wrapper for a file input stream that provides a subset of the functionality
 * of the java.io.DataInputStream class.
 */
typedef struct {
    ByteAddress in;  // the byte array
    UWord size;      // the size of the byte array
    UWord pos;       // the current read position
} DataInputStream;

void new_DataInputStream_open(DataInputStream *dis, unsigned char * fileArray, unsigned int size) {
	(*dis).size = size;
	if ((*dis).size == -1) {
//			fprintf(stderr, "No such file '%s'\n", fileArray);
			stopVM(-1);
	}

	//(*dis).in = newBuffer((*dis).size, "DataInputStream_open", true);
	//strcpy(fileArray, (*dis).in);
	//strncpy((*dis).in, fileArray, size);
	(*dis).in = (ByteAddress)fileArray;
	//snprintf((*dis).in, size, "%s", fileArray);
	(*dis).pos = 0;
}
/**
 * Initializes a DataInputStream.
 *
 * @param  dis   the DataInputStream to initialize
 * @param  file  the name of the file to initialize the stream from
 */
void DataInputStream_open(DataInputStream *dis, const char *file) {
	(*dis).size = getFileSize(file);
	if ((*dis).size == -1) {
		fprintf(stderr, "No such file '%s'\n", file);
		stopVM(-1);
	}

	// Note: the default way to allocate memory written by zeni
	(*dis).in = newBuffer((*dis).size, "DataInputStream_open", true);

	// Note: modified way to allocate memory written by zeni
	//(*dis).in = memget(sizeof(char)*(*dis).size);

	if (readFile(file, (*dis).in, (*dis).size) != (*dis).size) {
		printf("Error reading '%s'\n", file);
		stopVM(-1);
	}
	(*dis).pos = 0;
}

/**
 * Returns the number of bytes that can still be read from a DataInputStream stream.
 *
 * @param  dis   the DataInputStream to query
 * @return the number of available bytes
 */
unsigned int DataInputStream_available(DataInputStream *dis) {
	return (*dis).size - (*dis).pos;
}

/**
 * Returns the number of bytes that have been read from a DataInputStream stream.
 *
 * @param  dis   the DataInputStream to query
 * @return the number of bytes that have been read
 */
unsigned int DataInputStream_readSoFar(DataInputStream *dis) {
	return (*dis).pos;
}

/**
 * Reads an unsigned byte from a DataInputStream.
 *
 * @param  dis    the DataInputStream to read from
 * @param  prefix the prefix used when tracing this read
 * @return the value read
 */
UWord DataInputStream_readUnsignedByte(DataInputStream *dis, const char *prefix) {
	UWord value;

	if ((*dis).pos >= (*dis).size) {
		fatalVMError("EOFException");
	}
	value = (*dis).in[(*dis).pos++];
#if TRACE_SUITE
	if (prefix != null) {
		fprintf(stderr, "%s:%u\n", prefix, (value & 0xFF));
	}
#endif /* TRACE_SUITE */
	return value & 0xFF;
}

/**
 * Reads an unsigned short from a DataInputStream.
 *
 * @param  dis    the DataInputStream to read from
 * @param  prefix the prefix used when tracing this read
 * @return the value read
 */
UWord DataInputStream_readUnsignedShort(DataInputStream *dis,
		const char *prefix) {
	int b1 = DataInputStream_readUnsignedByte(dis, null);
	int b2 = DataInputStream_readUnsignedByte(dis, null);
	UWord value = ((b1 << 8) + (b2 << 0)) & 0xFFFF;
#if TRACE_SUITE
	fprintf(stderr, "%s:%u\n", prefix, value);
#endif /* TRACE_SUITE */
	return value;
}

/**
 * Reads an int from a DataInputStream.
 *
 * @param  dis    the DataInputStream to read from
 * @param  prefix the prefix used when tracing this read
 * @return the value read
 */
int DataInputStream_readInt(DataInputStream *dis, const char *prefix) {
	int b1 = DataInputStream_readUnsignedByte(dis, null);
	int b2 = DataInputStream_readUnsignedByte(dis, null);
	int b3 = DataInputStream_readUnsignedByte(dis, null);
	int b4 = DataInputStream_readUnsignedByte(dis, null);
	UWord value = ((b1 << 24) + (b2 << 16) + (b3 << 8) + (b4 << 0));
#if TRACE_SUITE
	fprintf(stderr, "%s:%u\n", prefix, value);
#endif /* TRACE_SUITE */
	return value;
}

/**
 * Reads some bytes from a DataInputStream into a buffer.
 *
 * @param  dis    the DataInputStream to read from
 * @param  buf    the buffer to store into
 * @param  size   the size of the buffer which also specifies the number of bytes to read
 * @param  prefix the prefix used when tracing this read
 * @return the value read
 */
void DataInputStream_readFully(DataInputStream *dis, ByteAddress buf,
		UWord size, const char *prefix) {
	unsigned int i;
	for (i = 0; i != size; ++i) {
//	while (i != size) {
		buf[i] = DataInputStream_readUnsignedByte(dis, null);
	}
#if TRACE_SUITE
	fprintf(stderr, format("%s:{read %A bytes}\n"), prefix, size);
#endif /* TRACE_SUITE */
}

/**
 * Skips a number of bytes in a DataInputStream.
 *
 * @param  dis    the DataInputStream to operate on
 * @param  n      the number of bytes to skip
 * @param  prefix the prefix used when tracing this read
 */
void DataInputStream_skip(DataInputStream *dis, UWord n, const char* prefix) {
	unsigned int i = 0;
	while (i++ != n) {
		DataInputStream_readUnsignedByte(dis, null);
	}
#if TRACE_SUITE
	fprintf(stderr, format("%s:{skipped %A bytes}\n"), prefix, n);
#endif /* TRACE_SUITE */
}

/**
 * Closes a DataInputStream.
 *
 * @param dis  the DataInputStream to close
 */
void DataInputStream_close(DataInputStream *dis) {
	if ((*dis).in != null) {
		// written by zeni
		freeBuffer((*dis).in);
		(*dis).in = null;
	}
}

void new_DataInputStream_close(DataInputStream *dis) {
	if ((*dis).in != null) {
		// written by zeni
		//freeBuffer((*dis).in);
		(*dis).in = null;
	}
}

// new loadBootstrapSuite
UWord new_loadBootstrapSuite(unsigned char * fileArray, unsigned int fileSize, Address buffer, UWord size,
		Address *suite, int *hash) {
	DataInputStream dis;
	UWord suiteOffset;
	ByteAddress oopMap;
	UWord oopMapLength;
	unsigned int i;
	int attributes;
	boolean hasTypemap;
	int pad/*, fileSize*/;
	if (fileSize == -1) {
		fprintf(stderr, "Bootstrap suite file '%s' not found\n", fileArray);
		fatalVMError("");
	} else if (size < fileSize) {
		fatalVMError("buffer size is too small for bootstrap suite");
	}
	/*
	 * Try to open the suite file.
	 */
	new_DataInputStream_open(&dis, fileArray, fileSize);
	/*
	 * Read 'magic'
	 */
	if (DataInputStream_readInt(&dis, "magic") != 0xdeadbeef) {
		fatalVMError("bad magic in bootstrap suite");
	}

	/*
	 * Read (and ignore for now) version identifiers
	 */
	DataInputStream_readUnsignedShort(&dis, "minor_version");
	DataInputStream_readUnsignedShort(&dis, "major_version");

	/*
	 * Read 'attributes'
	 */
	attributes = DataInputStream_readInt(&dis, "attributes");


	hasTypemap = ((attributes
			& com_sun_squawk_ObjectMemoryFile_ATTRIBUTE_TYPEMAP) != 0);
	if (((attributes & com_sun_squawk_ObjectMemoryFile_ATTRIBUTE_32BIT) != 0)
			== SQUAWK_64) {
		fatalVMError("word size in bootstrap suite is incorrect");
	}
	if (((attributes & com_sun_squawk_ObjectMemoryFile_ATTRIBUTE_BIGENDIAN) != 0)
			!= PLATFORM_BIG_ENDIAN) {
		fatalVMError("endianess in bootstrap suite is incorrect");
	}
	/*
	 * Read and ignore 'parent_hash'
	 */
	DataInputStream_readInt(&dis, "parent_hash");
	/*
	 * Read the length of 'parent_url' and verify that it is 0
	 */
	if (DataInputStream_readUnsignedShort(&dis, "parent_url") != 0) {
		fatalVMError("bootstrap suite should have no parent");
	}
	/*
	 * Read 'root'
	 */
	suiteOffset = DataInputStream_readInt(&dis, "root");
	/*
	 * Read 'size'
	 */
	size = DataInputStream_readInt(&dis, "size");
	/*
	 * Read 'oopmap'
	 */
	oopMapLength = ((size / HDR_BYTES_PER_WORD) + 7) / 8;
	oopMap = (ByteAddress) buffer + size;
//	oopMap = dis.in + dis.pos;
//	dis.pos += oopMapLength;
	DataInputStream_readFully(&dis, oopMap, oopMapLength, "oopmap");
	/*
	 * Skip the padding for 'memory' to be word aligned
	 */
	pad = roundUpToWord(DataInputStream_readSoFar(&dis))
			- DataInputStream_readSoFar(&dis);
	DataInputStream_skip(&dis, pad, "padding");
//	dis.pos += pad;
	/*
	 * Read 'memory'
	 */
	DataInputStream_readFully(&dis, buffer, size, "memory");
//	buffer = dis.in + dis.pos;
//	dis.pos += size;
	/*
	 * Calculate the hash of the object memory while it is still in canonical form.
	 */
	*hash = size;
	for (i = 0; i != size; ++i) {
		setType((ByteAddress) buffer + i, AddressType_BYTE, 1);
		*hash += getByte(buffer, i);
	}

	/*
	 * Relocate all the pointers in the object memory.
	 */
#if TRACE_SUITE
	{
		FILE *pointers = fopen("squawk.suite.pointers", "w");
		int cardinality = 0;
#endif /* TRACE_SUITE */

	for (i = 0; i != oopMapLength; ++i) {
		int oopMapByte = oopMap[i];
		int bit;

		for (bit = 0; bit != 8; ++bit) {
			if ((oopMapByte & (1 << bit)) != 0) {
				int offset = (i * 8) + bit;
				UWord pointer;
				setType((UWordAddress) buffer + offset, AddressType_UWORD,
						HDR_BYTES_PER_WORD);
				pointer = getUWord(buffer, offset);
				if (pointer != 0) {
					setUWord(buffer, offset, pointer + (UWord) buffer);
				}
#if TRACE_SUITE
				++cardinality;
				fprintf(pointers, format("%A: %A -> %A\n"), offset*HDR_BYTES_PER_WORD, pointer, pointer + (UWord)buffer);
#endif /* TRACE_SUITE */
			}
		}
	}
#if TRACE_SUITE
	fprintf(stderr, "oopmap:{cardinality = %d}\n", cardinality);
	fclose(pointers);
}
#endif /* TRACE_SUITE */
	/*
	 * Initialize the suite pointer.
	 */
	*suite = (ByteAddress) buffer + suiteOffset;

#if TYPEMAP
	if (hasTypemap) {
		ByteAddress p = buffer;
		int i;

		for (i = 0; i != size; ++i) {
			char type = DataInputStream_readUnsignedByte(&dis, "type");
			setType(p, type, 1);
			p++;
		}
	}
#endif /* TYPEMAP */
	/*
	 * Close the data input stream.
	 */
	new_DataInputStream_close(&dis);
	return size;
}
/**
 * Loads the bootstrap suite from a file into a given buffer.
 *
 * @param file    the name of the file to load from
 * @param buffer  the buffer into which the object memory is to be loaded
 * @param size    the size of the buffer. This must at least equal to the size of the file
 * @param suite   OUT: the pointer to the suite
 * @param hash    OUT: the hash of the object memory in canonical form
 * @return the size of the object memory
 */
UWord loadBootstrapSuite(const char *file, Address buffer, UWord size,
		Address *suite, int *hash) {
	DataInputStream dis;
	UWord suiteOffset;
	ByteAddress oopMap;
	UWord oopMapLength;
	unsigned int i;
	int attributes;
	boolean hasTypemap;
	int pad, fileSize;

	fileSize = getFileSize(file);
	if (fileSize == -1) {
		fprintf(stderr, "Bootstrap suite file '%s' not found\n", file);
		fatalVMError("");
	} else if (size < fileSize) {
		fatalVMError("buffer size is too small for bootstrap suite");
	}
	/*
	 * Try to open the suite file.
	 */
	DataInputStream_open(&dis, file);

	/*
	 * Read 'magic'
	 */
	if (DataInputStream_readInt(&dis, "magic") != 0xdeadbeef) {
		fatalVMError("bad magic in bootstrap suite");
	}

	/*
	 * Read (and ignore for now) version identifiers
	 */
	DataInputStream_readUnsignedShort(&dis, "minor_version");
	DataInputStream_readUnsignedShort(&dis, "major_version");

	/*
	 * Read 'attributes'
	 */
	attributes = DataInputStream_readInt(&dis, "attributes");
	hasTypemap = ((attributes
			& com_sun_squawk_ObjectMemoryFile_ATTRIBUTE_TYPEMAP) != 0);
	if (((attributes & com_sun_squawk_ObjectMemoryFile_ATTRIBUTE_32BIT) != 0)
			== SQUAWK_64) {
		fatalVMError("word size in bootstrap suite is incorrect");
	}
	if (((attributes & com_sun_squawk_ObjectMemoryFile_ATTRIBUTE_BIGENDIAN) != 0)
			!= PLATFORM_BIG_ENDIAN) {
		fatalVMError("endianess in bootstrap suite is incorrect");
	}
	/*
	 * Read and ignore 'parent_hash'
	 */
	DataInputStream_readInt(&dis, "parent_hash");
	/*
	 * Read the length of 'parent_url' and verify that it is 0
	 */
	if (DataInputStream_readUnsignedShort(&dis, "parent_url") != 0) {
		fatalVMError("bootstrap suite should have no parent");
	}
	/*
	 * Read 'root'
	 */
	suiteOffset = DataInputStream_readInt(&dis, "root");
	/*
	 * Read 'size'
	 */
	size = DataInputStream_readInt(&dis, "size");
	/*
	 * Read 'oopmap'
	 */
	oopMapLength = ((size / HDR_BYTES_PER_WORD) + 7) / 8;
	oopMap = (ByteAddress) buffer + size;

	DataInputStream_readFully(&dis, oopMap, oopMapLength, "oopmap");
	/*
	 * Skip the padding for 'memory' to be word aligned
	 */
	pad = roundUpToWord(DataInputStream_readSoFar(&dis))
			- DataInputStream_readSoFar(&dis);
	DataInputStream_skip(&dis, pad, "padding");
	/*
	 * Read 'memory'
	 */
	DataInputStream_readFully(&dis, buffer, size, "memory");
	/*
	 * Calculate the hash of the object memory while it is still in canonical form.
	 */
	*hash = size;
	for (i = 0; i != size; ++i) {
		setType((ByteAddress) buffer + i, AddressType_BYTE, 1);
		*hash += getByte(buffer, i);
	}

	/*
	 * Relocate all the pointers in the object memory.
	 */
#if TRACE_SUITE
	{
		FILE *pointers = fopen("squawk.suite.pointers", "w");
		int cardinality = 0;
#endif /* TRACE_SUITE */

	for (i = 0; i != oopMapLength; ++i) {
		int oopMapByte = oopMap[i];
		int bit;

		for (bit = 0; bit != 8; ++bit) {
			if ((oopMapByte & (1 << bit)) != 0) {
				int offset = (i * 8) + bit;
				UWord pointer;
				setType((UWordAddress) buffer + offset, AddressType_UWORD,
						HDR_BYTES_PER_WORD);
				pointer = getUWord(buffer, offset);
				if (pointer != 0) {
					setUWord(buffer, offset, pointer + (UWord) buffer);
				}
#if TRACE_SUITE
				++cardinality;
				fprintf(pointers, format("%A: %A -> %A\n"), offset*HDR_BYTES_PER_WORD, pointer, pointer + (UWord)buffer);
#endif /* TRACE_SUITE */
			}
		}
	}

#if TRACE_SUITE
	fprintf(stderr, "oopmap:{cardinality = %d}\n", cardinality);
	fclose(pointers);
}
#endif /* TRACE_SUITE */

	/*
	 * Initialize the suite pointer.
	 */
	*suite = (ByteAddress) buffer + suiteOffset;

#if TYPEMAP
	if (hasTypemap) {
		ByteAddress p = buffer;
		int i;

		for (i = 0; i != size; ++i) {
			char type = DataInputStream_readUnsignedByte(&dis, "type");
			setType(p, type, 1);
			p++;
		}
	}
#endif /* TYPEMAP */
	/*
	 * Close the data input stream.
	 */
	DataInputStream_close(&dis);
	return size;
}

#endif /* FLASH_MEMORY */
